### PR TITLE
Add benchmarks to measure Dictionary.filter

### PR DESCRIPTION
**Pull request by rctcwyvrn**
Adds benchmarks to measure the effect of https://github.com/apple/swift/pull/59243

### CODE REVIEW

**Comment by rctcwyvrn** [comment]
@swift-ci benchmark

**Comment by rctcwyvrn** [comment]
@swift-ci smoke test

File: benchmark/single-source/DictionaryRemove.swift [hunk 1]
```diff
@@ -26,6 +26,12 @@ public let benchmarks = [
     runFunction: remove, tags: t, legacyFactor: 10),
   BenchmarkInfo(name: "DictionaryRemoveOfObjects",
     runFunction: removeObjects, tags: t, legacyFactor: 100),
+
+  BenchmarkInfo(name: "DictionaryFilter",
+    runFunction: filter, tags: t, legacyFactor: 10),
```
Comment (lorentey) [inline benchmark/single-source/DictionaryRemove.swift:31]: The legacy factor is only needed for older benchmark functions that we (used to?) not be able to change without breaking our ability to compare benchmark results across compiler releases. We should not use it for new definitions -- instead, I recommend just incorporating this into the `100*n` factors below. (iirc, changing it to `1000*n` should work. (Or is it `10*n`?))

File: benchmark/single-source/DictionaryRemove.swift [hunk 2]
```diff
@@ -59,3 +65,19 @@ func removeObjects(n: Int) {
     check(dict.isEmpty)
   }
 }
+
+func filter(n: Int) {
+  for _ in 1...100*n {
+    let dict = numberMap
```
Comment (lorentey) [inline benchmark/single-source/DictionaryRemove.swift:71]: The first invocation of `numberMap` (and `boxedNumMap` below) is going to initialize the dictionary value at runtime, which could lead to some instability here. Let's add a `setUpFunction` to the `BenchmarkInfo` declaration that simply accesses its getter to make sure that happens outside of the actual benchmarking run. Something like `setUpFunction: { blackHole(numberMap) }` ought to do the trick! (We should add this for the existing two benchmarks, too.)

**Review by lorentey — APPROVED** [review]
Looks great, thank you! 👍

**Comment by rctcwyvrn** [comment]
@swift-ci smoke test

**Comment by rctcwyvrn** [comment]
@swift-ci benchmark

**Pull request merged**